### PR TITLE
Add end-to-end PCPhase support

### PR DESCRIPTION
**Pull request by maliasadi**
**Context:**
LQ and LGPU provide native support for PCPhase. To enable this in Catalyst, we had to define a new custom MLIR gate op in the Quantum dialect and extend both the frontend and runtime to lower PCPhase into a new QIR instruction. This allows execution on devices with native support and unlocks simulation of QSVT workloads w/ ~30 qubits on a single GPU.

**Description of the Change:**

**Benefits:**
- End-to-end PCPhase support
- Simulate QSVT workflows with ~30 qubits backed by LGPU

**Possible Drawbacks:**
- We need a better representation in Catalyst to support generic PL gates with hyperparameters.

**Related GitHub Issues:**
[sc-98149]

### CODE REVIEW

File: benchmark/batchrun.py [hunk 1]
```diff
@@ -14,15 +14,7 @@
 """Measurement cycle management entry point"""
 import sys
 
-from catalyst_benchmark.toplevel import (
```
Comment (maliasadi) [inline benchmark/batchrun.py:17]: `make format` updated these headers

File: runtime/tests/Test_NullQubit.cpp [hunk 2]
```diff
@@ -353,7 +353,7 @@ TEST_CASE("Test NullQubit state vector after AllocateQubits", "[NullQubit]")
 
     CHECK(sim->AllocateQubits(0).size() == 0);
 
-    auto &&q = sim->AllocateQubits(2);
+    sim->AllocateQubits(2);
```
Comment (maliasadi) [inline runtime/tests/Test_NullQubit.cpp:356]: It's an unused variable. `make runtime-test` fails without this change.

File: mlir/include/Quantum/IR/QuantumOps.td [hunk 3]
```diff
@@ -637,10 +637,60 @@ def MultiRZOp : UnitaryGate_Op<"multirz", [DifferentiableGate, NoMemoryEffect,
             return getODSOperands(getParamOperandIdx());
         }
     }];
-    
+
+    let hasCanonicalizeMethod = 1;
+}
+
+def PCPhaseOp : UnitaryGate_Op<"pcphase", [DifferentiableGate, NoMemoryEffect,
```
Comment (paul0403) [inline mlir/include/Quantum/IR/QuantumOps.td:644]: Hey, I think this looks good and is relatively straightforward, just a question: why did you decide to add it as its own new op, instead of subsuming it under custom ops?

File: frontend/catalyst/device/qjit_device.py [hunk 4]
```diff
@@ -76,6 +76,7 @@
     "PauliX",
     "PauliY",
     "PauliZ",
+    "PCPhase",
```
Comment (paul0403) [inline frontend/catalyst/device/qjit_device.py:81]: Remember to add a changelog!

File: runtime/lib/capi/RuntimeCAPI.cpp [hunk 5]
```diff
@@ -504,6 +504,29 @@ void __catalyst__qis__SetState(MemRefT_CplxT_double_1d *data, uint64_t numQubits
...
+void __catalyst__qis__PCPhase(double theta, double dim, const Modifiers *modifiers,
+                              int64_t numQubits, ...)
+{
+    RT_ASSERT(numQubits >= 0);
+
+    // Compute the abs value to support adjoint
+    // Note this will be revisited once we have a better solution
+    // for capturing generic PL gates. Currently, this is the optimal
+    // workaround to reduce type casting throughout the pipeline.
+    dim = dim < 0 ? -dim : dim;
```
Comment (paul0403) [inline runtime/lib/capi/RuntimeCAPI.cpp]: Ah, is this why PCPhase needs to be its own op? Because its runtime needs its specific preprocessing before going into the `NamedOperation`?

(Also, why can the dimension be a negative float? :sweat_smile: )

File: mlir/include/Quantum/IR/QuantumOps.td (hunk 3 above)
Comment (maliasadi) [reply mlir/include/Quantum/IR/QuantumOps.td:644]: Mainly because of differences in the signature of `PCPhase`; this op works with variadic number of wires and has the `dim` hyperparameter. I followed the same decision made for `GlobalPhase` and `MultiRZ`, and implement it as a new op instead of trying to update the legacy custom op.

File: runtime/lib/capi/RuntimeCAPI.cpp (hunk 5 above)
Comment (maliasadi) [reply runtime/lib/capi/RuntimeCAPI.cpp]: Yeah, `NamedOperation` in Catalyst Runtime and `ApplyOperation` in Lightning C++ APIs don't accept hyperparameters  as the rsult we need to serialize them as regular params before calling the gate kernels. Here for PCPhase, negative dim is not accepted by [PL](https://docs.pennylane.ai/en/stable/code/api/pennylane.PCPhase.html) and Lightning. When applying adjiont to the op, the list of params will be multiplied by negative one, this little patch here is needed to fix this.

I wanted to ensure that we have `dim` in the assembly of PCPhase while reducing the number of type casts to one place so that we can fix this later in future when we have a better support for Operator's hyperparamters in Catalyst.

File: doc/releases/changelog-dev.md [hunk 6]
```diff
@@ -76,6 +76,9 @@
 
 <h3>Internal changes ⚙️</h3>
 
+* Add `PCPhase` native compiler and runtime operator support.
+  [(#1987)](https://github.com/PennyLaneAI/catalyst/pull/1987)
+
 * Add `SingleExcitation` and `DoubleExcitation` to native `RUNTIME_OPERATIONS` and `RuntimeCAPI`.
   [(#1980)](https://github.com/PennyLaneAI/catalyst/pull/1980)
```
Comment (dime10) [inline doc/releases/changelog-dev.md]: Aren't these user-facing changes? They can now use these gates without decomposition in their programs. We should also mention that this significantly speeds up simulation runtime in the changelog (the "so what").

File: mlir/include/Quantum/IR/QuantumOps.td (hunk 3 above)
Comment (dime10) [reply mlir/include/Quantum/IR/QuantumOps.td:644]: I think @rniczh recently unified/simplified the lowering of variadic qubit ops?

File: mlir/include/Quantum/IR/QuantumOps.td [hunk 7]
```diff
@@ -637,10 +637,60 @@ def MultiRZOp : UnitaryGate_Op<"multirz", [DifferentiableGate, NoMemoryEffect,
             return getODSOperands(getParamOperandIdx());
         }
     }];
-    
+
+    let hasCanonicalizeMethod = 1;
```
Comment (paul0403) [inline mlir/include/Quantum/IR/QuantumOps.td:641]: What is this canonicalization method on multirz about?

File: mlir/include/Quantum/IR/QuantumOps.td (hunk 3 above)
Comment (paul0403) [reply mlir/include/Quantum/IR/QuantumOps.td:644]: > I think @rniczh recently unified/simplified the lowering of variadic qubit ops?

Only in the custom op lowering patterns, i.e. only in the sense of `quantum.custom "gate_with_variadic_inqubits"` can be lowered to `__catalyst__qis__gate_with_variadic_inqubits(...)` through the `CustomOpPattern` https://github.com/PennyLaneAI/catalyst/blob/main/mlir/lib/Quantum/Transforms/ConversionPatterns.cpp#L443. The underlying `__catalyst__qis__pcphase` still needs to exist. 

Hmm, actually now as I am typing this, I realized maybe we can go through custom op for pcphase as well? i.e.

1. During binding `qinst_p`, pcphase still has the special treatement for `dim`:
```python
jax_tracer.py
            params = (
                op.parameters + [op.hyperparameters["dimension"][0]]
                if isinstance(op, qml.PCPhase)
                else op.parameters
            )
            qinst_p.bind(...params...)
``` 

2. During lowering jaxpr to mlir, just treat `dim` as another param in `float_params` on custom op
```python
jax_primitives.py
        float_param = float_params[0]
        dim_param = float_params[1]
        return CustomOp(
            out_qubits=[qubit.type for qubit in qubits],
            out_ctrl_qubits=[qubit.type for qubit in ctrl_qubits],
            theta=float_param,
            dim=dim_param,
            in_qubits=qubits,
            in_ctrl_qubits=ctrl_qubits,
            in_ctrl_values=ctrl_values_i1,
            adjoint=adjoint,
        ).results
```

3. During custom op lowering to runtime, add pcphase as a variadic op https://github.com/PennyLaneAI/catalyst/blob/main/mlir/lib/Quantum/Transforms/ConversionPatterns.cpp#L443
```
Quantum/ConversionPatterns.cpp
      const auto hasVariadicInputQbits = (op.getGateName() == "Identity") || (op.getGateName() == "PCPhase"));
```

and this would lower to your new `__catalyst__qis__PCPhase` runtime stub.

@maliasadi so I think going through custom op is still possible. But I don't really have a preference either way, so I'm fine with whichever method is easier for you. @dime10 David do you have a preference here?

Comment (dime10) [reply mlir/include/Quantum/IR/QuantumOps.td:644]: It's already a separate op, so I don't mind tbh (using custom is also fine though). Who knows maybe at some point we will have a separate operation for each known gate 😅 

However, make sure all new operations implement all applicable quantum interfaces!

File: doc/releases/changelog-dev.md (hunk 6 above)
Comment (maliasadi) [reply doc/releases/changelog-dev.md]: yeah, I was thinking the same. I'm benchmarking QSVT w/ this PR for the presentation. I'll update the changelog adding more details about the "so what" :smile:

File: mlir/include/Quantum/IR/QuantumOps.td (hunk 3 above)
Comment (maliasadi) [reply mlir/include/Quantum/IR/QuantumOps.td:644]: > Who knows maybe at some point we will have a separate operation for each known gate

I’m kinda leaning toward this idea, honestly

Comment (maliasadi) [reply mlir/include/Quantum/IR/QuantumOps.td:644]: @paul0403 @dime10 Let me know what's the preferred approach so that I can finalize and merge this one today or tomorrow.

Comment (dime10) [reply mlir/include/Quantum/IR/QuantumOps.td:644]: Up to you, as long as this is respected:
> However, make sure all new operations implement all applicable quantum interfaces!

Comment (dime10) [reply mlir/include/Quantum/IR/QuantumOps.td:644]: I you don't care I would just say use the new op 👍

File: mlir/lib/Quantum/IR/QuantumOps.cpp [hunk 8]
```diff
@@ -84,6 +84,21 @@ LogicalResult MultiRZOp::canonicalize(MultiRZOp op, mlir::PatternRewriter &rewri
     return failure();
 }
 
+LogicalResult PCPhaseOp::canonicalize(PCPhaseOp op, mlir::PatternRewriter &rewriter)
+{
+    if (op.getAdjoint()) {
+        auto paramNeg = rewriter.create<mlir::arith::NegFOp>(op.getLoc(), op.getTheta());
+        auto paramDim = rewriter.create<mlir::arith::NegFOp>(op.getLoc(), op.getDim());
```
Comment (paul0403) [inline mlir/lib/Quantum/IR/QuantumOps.cpp]: Maybe we can simply not negate the dimension here if adjoint is present? This would avoid the need to absolute value the dim in the runtime stub right?

File: mlir/lib/Quantum/Transforms/merge_rotation.cpp [hunk 9]
```diff
@@ -53,6 +53,8 @@ struct MergeRotationsPass : impl::MergeRotationsPassBase<MergeRotationsPass> {
                                                                  &getContext());
         catalyst::quantum::MultiRZOp::getCanonicalizationPatterns(patternsCanonicalization,
                                                                   &getContext());
+        catalyst::quantum::PCPhaseOp::getCanonicalizationPatterns(patternsCanonicalization,
```
Comment (paul0403) [inline mlir/lib/Quantum/Transforms/merge_rotation.cpp]: Do we care about PCPhase in merge-rotation pass? I feel like that would require some extra work (i.e. check the merged pcphase dimensions are the same, etc.). I propose we leave PCPhase's merge rotation as a separate issue (which would also be a good first issue for assessment/onboarding/...)

File: mlir/test/Quantum/ConversionTest.mlir [hunk 10]
```diff
@@ -331,6 +331,35 @@ func.func @multirz(%q0 : !quantum.bit, %p : f64) -> (!quantum.bit, !quantum.bit,
...
+
+    // CHECK: [[d:%.+]] = llvm.mlir.zero : !llvm.ptr
+    // CHECK: [[c1:%.+]] = llvm.mlir.constant(1 : i64)
+    // CHECK: llvm.call @__catalyst__qis__PCPhase(%arg1, %arg2, [[d]], [[c1]], %arg0)
+    %q1 = quantum.pcphase(%p, %d) %q0 : !quantum.bit
+
+    // CHECK: [[d:%.+]] = llvm.mlir.zero : !llvm.ptr
+    // CHECK: [[c2:%.+]] = llvm.mlir.constant(2 : i64)
+    // CHECK: llvm.call @__catalyst__qis__PCPhase(%arg1, %arg2, [[d]], [[c2]], %arg0, %arg0)
+    %q2:2 = quantum.pcphase(%p, %d) %q1, %q1 : !quantum.bit, !quantum.bit
```
Comment (paul0403) [inline mlir/test/Quantum/ConversionTest.mlir:347]: Why did the second pcphase take in `%arg0` again? Shouldn't it take in `%q1`, the out qubits of the first pcphase? :scream: :thinking:

Comment (paul0403) [reply mlir/test/Quantum/ConversionTest.mlir:347]: Ah this is runtime IR now, so they both apply to the same qubit that has been getelemntptr-ed. My bad!

File: runtime/lib/backend/common/Utils.hpp [hunk 11]
```diff
@@ -220,6 +220,7 @@ enum class SimulatorGate : uint8_t {
     Toffoli,
     // n-qubit
     MultiRZ,
+    PCPhase,
```
Comment (paul0403) [inline runtime/lib/backend/common/Utils.hpp]: What are the two lists of gates in `Utils.hpp` for? I did not see them for Joseph's single/double excitation gates :thinking:

File: runtime/lib/capi/RuntimeCAPI.cpp [hunk 12]
```diff
@@ -504,6 +504,29 @@ void __catalyst__qis__SetState(MemRefT_CplxT_double_1d *data, uint64_t numQubits
     getQuantumDevicePtr()->SetState(data_view, wires);
 }
 
+void __catalyst__qis__PCPhase(double theta, double dim, const Modifiers *modifiers,
+                              int64_t numQubits, ...)
+{
+    RT_ASSERT(numQubits >= 0);
+
```
Comment (paul0403) [inline runtime/lib/capi/RuntimeCAPI.cpp]: Do we want to assert that `dim` here is a integer? e.g. floor(dim) == dim or something?

File: runtime/tests/Test_NullQubit.cpp [hunk 13]
```diff
@@ -353,7 +353,7 @@ TEST_CASE("Test NullQubit state vector after AllocateQubits", "[NullQubit]")
...
 
     CHECK(sim->GetNumQubits() == 2);
```
Comment (paul0403) [inline runtime/tests/Test_NullQubit.cpp:358]: Maybe we can add a runtime test with null qubit :wink:

File: mlir/include/Quantum/IR/QuantumOps.td (hunk 7 above)
Comment (maliasadi) [reply mlir/include/Quantum/IR/QuantumOps.td:641]: To negate the param when adjoint is provided.

File: mlir/lib/Quantum/IR/QuantumOps.cpp (hunk 8 above)
Comment (maliasadi) [reply mlir/lib/Quantum/IR/QuantumOps.cpp]: hahaha... good catch! :smile:

File: mlir/lib/Quantum/Transforms/merge_rotation.cpp (hunk 9 above)
Comment (maliasadi) [reply mlir/lib/Quantum/Transforms/merge_rotation.cpp]: PCPhase is a non-global version of GlobalPhase w/ controls. We should be able to write a merge rotation transform for it. But I agree this takes a bit of time. I removed the draft impl but forgot this one. Removed it [ad9b147](https://github.com/PennyLaneAI/catalyst/pull/1987/commits/ad9b14740d5f2e019f26488b837750d562e9f334)

File: runtime/lib/backend/common/Utils.hpp (hunk 11 above)
Comment (maliasadi) [reply runtime/lib/backend/common/Utils.hpp]: Akh! old habit :smile: we used to use this in all the device classes. This is still being used in `OpenQasmDevice`. I never liked these lists btw! I implemented this fast and static `gate_lookup` in the first version of the runtime but never revisited it since :melting_face:

File: runtime/tests/Test_NullQubit.cpp (hunk 13 above)
Comment (maliasadi) [reply runtime/tests/Test_NullQubit.cpp:358]: Although we don't tests all these gates.. it's a good idea! Added [ad9b147](https://github.com/PennyLaneAI/catalyst/pull/1987/commits/ad9b14740d5f2e019f26488b837750d562e9f334)

File: runtime/lib/capi/RuntimeCAPI.cpp (hunk 12 above)
Comment (maliasadi) [reply runtime/lib/capi/RuntimeCAPI.cpp]: Yeah, this can be fixed now after updating the adjoint in canonicalization [ad9b147](https://github.com/PennyLaneAI/catalyst/pull/1987/commits/ad9b14740d5f2e019f26488b837750d562e9f334)

File: doc/releases/changelog-dev.md (hunk 6 above)
Comment (maliasadi) [reply doc/releases/changelog-dev.md]: Updated [10a0616](https://github.com/PennyLaneAI/catalyst/pull/1987/commits/10a06168edab2f1425e3f952823cc31c74806327)

Comment (maliasadi) [reply doc/releases/changelog-dev.md]: benchmark script is still running. I'll share the results later

File: mlir/include/Quantum/IR/QuantumOps.td (hunk 7 above)
Comment (paul0403) [reply mlir/include/Quantum/IR/QuantumOps.td:641]: Ah, so multirz always had a canonicalization pattern but its `let hasCanonicalizeMethod = 1;` was never set?

Comment (maliasadi) [reply mlir/include/Quantum/IR/QuantumOps.td:641]: hmm... it's more like a GH rendering issue :smile: I've added this for PCPhase

Comment (paul0403) [reply mlir/include/Quantum/IR/QuantumOps.td:641]: :exploding_head: I see it now
